File: data/cpp/tests/test_pycrypto_decrypt.cpp
#include <gtest/gtest.h>
#include "../src/function_pycrypto_decrypt.cpp"

class TestAESCipherProcess : public ::testing::Test {
protected:
    void SetUp() override {
        plaintext = "This is a test message for AES.";
        key = "thisis256bitkey______________";
    }

    std::string plaintext;
    std::string key;
};

TEST_F(TestAESCipherProcess, EncryptDecryptValidText) {
    std::string encrypted = AESCipher::process(key, plaintext, "encrypt");
    std::string decrypted = AESCipher::process(key, encrypted, "decrypt");
    EXPECT_EQ(decrypted, plaintext);
}

TEST_F(TestAESCipherProcess, DecryptInvalidEncryptedText) {
    std::string invalid_encrypted_text = "invalidtext";
    EXPECT_THROW(AESCipher::process(key, invalid_encrypted_text, "decrypt"), std::invalid_argument);
}

TEST_F(TestAESCipherProcess, EncryptDecryptEmptyString) {
    std::string encrypted_empty = AESCipher::process(key, "", "encrypt");
    std::string decrypted_empty = AESCipher::process(key, encrypted_empty, "decrypt");
    EXPECT_EQ(decrypted_empty, "");
}

TEST_F(TestAESCipherProcess, DecryptAfterDoubleEncryption) {
    std::string encrypted_once = AESCipher::process(key, plaintext, "encrypt");
    std::string decrypted_once = AESCipher::process(key, encrypted_once, "decrypt");
    EXPECT_EQ(decrypted_once, plaintext);

    std::string encrypted_twice = AESCipher::process(key, decrypted_once, "encrypt");
    std::string decrypted_twice = AESCipher::process(key, encrypted_twice, "decrypt");
    EXPECT_EQ(decrypted_twice, plaintext);
}

TEST_F(TestAESCipherProcess, DecryptWithAlteredCiphertext) {
    std::string encrypted = AESCipher::process(key, plaintext, "encrypt");
    std::string altered_encrypted = encrypted.substr(0, encrypted.length() - 1) + "x";
    EXPECT_THROW(AESCipher::process(key, altered_encrypted, "decrypt"), std::invalid_argument);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: data/python/tests/test_pycrypto_decrypt.py
import unittest
from function_pycrypto_decrypt import AESCipher

class TestAESAESCipherProcess(unittest.TestCase):

    def setUp(self):
        self.plaintext = 'This is a test message for AES.'
        self.key = 'thisis256bitkey______________'

    def test_encrypt_decrypt_valid_text(self):
        encrypted = AESCipher.process(self.key, self.plaintext, mode='encrypt')
        decrypted = AESCipher.process(self.key, encrypted, mode='decrypt')
        self.assertEqual(decrypted, self.plaintext)

    def test_decrypt_invalid_encrypted_text(self):
        invalid_encrypted_text = 'invalidtext'
        with self.assertRaises(ValueError):
            AESCipher.process(self.key, invalid_encrypted_text, mode='decrypt')

    def test_encrypt_decrypt_empty_string(self):
        encrypted_empty = AESCipher.process(self.key, '', mode='encrypt')
        decrypted_empty = AESCipher.process(self.key, encrypted_empty, mode='decrypt')
        self.assertEqual(decrypted_empty, '')

    def test_decrypt_after_double_encryption(self):
        encrypted_once = AESCipher.process(self.key, self.plaintext, mode='encrypt')
        decrypted_once = AESCipher.process(self.key, encrypted_once, mode='decrypt')
        self.assertEqual(decrypted_once, self.plaintext)

        encrypted_twice = AESCipher.process(self.key, decrypted_once, mode='encrypt')
        decrypted_twice = AESCipher.process(self.key, encrypted_twice, mode='decrypt')
        self.assertEqual(decrypted_twice, self.plaintext)

    def test_decrypt_with_altered_ciphertext(self):
        encrypted = AESCipher.process(self.key, self.plaintext, mode='encrypt')
        altered_encrypted = encrypted[:-1] + 'x'
        with self.assertRaises(ValueError):
            AESCipher.process(self.key, altered_encrypted, mode='decrypt')

if __name__ == '__main__':
    unittest.main()
